### PR TITLE
Update format Lambda to generate GeoHarvester extract command

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### Purpose and background context

Update format Lambda function to generate GeoHarvester extract command based on `source` input variable provided in payload to TIMDEX StepFunction.

For more details, see Confluence documentation: [TIMDEX StepFunction Updates | Pipeline Lambdas](https://mitlibraries.atlassian.net/wiki/spaces/GDT/pages/3562274862/TIMDEX+StepFunction+Updates#Pipeline-Lambdas)

### How can a reviewer manually see the effects of these changes?

1. [Build the Container](https://github.com/MITLibraries/timdex-pipeline-lambdas/tree/GDT-115-generate-geoharvester-extract-command?tab=readme-ov-file#build-the-container)
2. [Run the default handler for the container](https://github.com/MITLibraries/timdex-pipeline-lambdas/tree/GDT-115-generate-geoharvester-extract-command?tab=readme-ov-file#run-the-default-handler-for-the-container)
3. Open a new terminal and post the following to the container:
   ```
   curl -XPOST "http://localhost:9000/2015-03-31/functions/function/invocations" -d '{"next-step": "extract", "run-date": "2022-03-10T16:30:23Z", "run-type": "daily", "source": "gismit", "verbose": "true"}'
   ```
4. Observe the GeoHarvester extract command in the following output:
   ```
   {"run-date": "2022-03-10", "run-type": "daily", "source": "gismit", "verbose": true, "harvester-type": "geo", "next-step": "transform", "extract": {"extract-command": ["--verbose", "harvest", "--harvest-type=incremental", "--from-date=2022-03-09", "--output-file=s3://timdex-bucket-name/gismit/gismit-2022-03-10-daily-extracted-records-to-index.jsonl", "mit"]}}
   ```
    * Notice that (a) `--harvest-type=incremental` because `run-type=daily` and the (b) `.jsonl` file type appended to the `output-file` parameter.
### Includes new or updated dependencies?
NO

### Changes expectations for external applications?
NO

### What are the relevant tickets?
https://mitlibraries.atlassian.net/browse/GDT-115

### Developer
- [x] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [x] All related Jira tickets are linked in commit message(s)
- [x] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [x] Any manual tests have been performed and verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: lambdas/helpers.py [hunk 1]
```diff
@@ -54,7 +55,7 @@ def generate_step_output_filename(
     """
     sequence_suffix = f"_{sequence}" if sequence else ""
     if step == "extract":
-        file_type = "xml"
+        file_type = "jsonl" if "gis" in source else "xml"
```
Comment (ghukill) [inline lambdas/helpers.py:58]: To avoid the risk of the string `"gis"` being present in another source name -- current or future -- I'd propose that we make this an explicit list like you have other places, e.g.

```python
if source in ["gismit", "gisogm"]:
    # logic here to set .jsonl file extension
```

File: tests/test_helpers.py [hunk 2]
```diff
@@ -30,14 +30,16 @@ def test_generate_index_name():
 
 def test_generate_step_output_filename_with_sequence():
     assert (
-        helpers.generate_step_output_filename("index", "prefix", "extract", "01")
+        helpers.generate_step_output_filename(
+            "source", "index", "prefix", "extract", "01"
+        )
         == "prefix-to-index_01.xml"
     )
```
Comment (ghukill) [inline tests/test_helpers.py:37]: This test is helpful that is confirms a `.xml` filename is created by default.  Would you be willing to add a test that confirms both `gismit` and `gisogm` sources get a `.jsonl` extension?

File: lambdas/commands.py [hunk 3]
```diff
@@ -20,32 +20,47 @@ def generate_extract_command(
...
+        extract_command.append(f"--metadata-format={input_data['oai-metadata-format']}")
+        if run_type == "daily":
+            extract_command.append(
+                f"--from-date={helpers.generate_harvest_from_date(run_date)}",
+            )
+        elif run_type == "full":
+            extract_command.append("--exclude-deleted")
+
+        if set_spec := input_data.get("oai-set-spec"):
+            extract_command.append(f"--set-spec={set_spec}")
```
Comment (ghukill) [inline lambdas/commands.py:63]: While I think these changes are very clean, easy to follow, and inline with the previous style of this function, what do you think about breaking the functionality of creating OAI and Geo harvester command generation into helper functions?  e.g. something like `generate_oai_harvester_extract_command()` and `generate_geo_harvester_extract_command()`?  They could live in the `helpers.py` file for the time being?

Ultimately, I'd imagine it would make them easier to test, though it would require some changes to the current tests.

File: lambdas/config.py [hunk 4]
```diff
@@ -83,14 +83,15 @@ def validate_input(input_data: dict) -> None:
...
-            raise ValueError(message)
+        if input_data["source"] not in ["gismit", "gisogm"]:
+            if missing_harvest_fields := [
+                field for field in REQUIRED_OAI_HARVEST_FIELDS if field not in input_data
+            ]:
+                message = (
+                    "Input must include all required harvest fields when starting with "
+                    f"harvest step. Missing fields: {missing_harvest_fields}"
+                )
+                raise ValueError(message)
```
Comment (ghukill) [inline lambdas/config.py:94]: While this definitely works, just noting a lot of nested `if` statements here.  However, because the OAI Harvester is the "default", you kind of do need a negation check here.

Ultimately, I could see validation functions for `next-step`, `run-type`, and then for `extract` it could hand off to validators depending on the harvester.  

However, not sure that's worth the effort now.  If we are going to explore unifying the TIMDEX sources across parts of the pipeline, where this lambda is a critical piece of that, might be better to leave as-is -- which is readable and easy to scan -- and consider refactoring down the road.

**Review by ghukill — CHANGES_REQUESTED** [review]
Overall, think it looks good.  I believe it would support the functionality needed.  

Requesting a couple changes and happy to discuss further.

File: lambdas/commands.py (hunk 3 above)
Comment (jonavellecuerdo) [reply lambdas/commands.py:63]: As discussed during our check-in, we decided to leave the code as-is for now and make additional improvements to the architecture (e.g., explore benefits of a source-level module) in the near future (potentially when we integrate `browsertrix-harvester` into TIMDEX).

File: lambdas/helpers.py (hunk 1 above)
Comment (jonavellecuerdo) [reply lambdas/helpers.py:58]: Updated!

File: lambdas/config.py (hunk 4 above)
Comment (jonavellecuerdo) [reply lambdas/config.py:94]: **Note:** Will leave this comment 'Unresolved' for easier location

File: tests/test_helpers.py (hunk 2 above)
Comment (jonavellecuerdo) [reply tests/test_helpers.py:48]: See latest commit!

**Comment by jonavellecuerdo** [comment]
@ghukill and @ehanson8 -- Please see the latest updates to this PR!

File: lambdas/format_input.py [hunk 5]
```diff
@@ -12,7 +11,6 @@ def lambda_handler(event: dict, _context: dict) -> dict:
     config.verify_env()
     verbose = config.check_verbosity(event.get("verbose", False))
     config.configure_logger(logging.getLogger(), verbose)
-    logger.debug(json.dumps(event))
```
Comment (ehanson8) [inline lambdas/format_input.py:15]: Why was this removed?

**Review by ehanson8 — APPROVED** [review]
Looks good and given the comments about your discussions with Graham, I think it's good to go. 1 question about why some logging was removed

File: lambdas/commands.py [hunk 6]
```diff
@@ -20,32 +20,47 @@ def generate_extract_command(
...
 
     if verbose:
         extract_command.append("--verbose")
 
-    extract_command.append("harvest")
-    extract_command.append(f"--metadata-format={input_data['oai-metadata-format']}")
-
-    if source in ["aspace", "dspace"]:
-        extract_command.append("--method=get")
+    if source in ["gismit", "gisogm"]:
```
Comment (ghukill) [inline lambdas/commands.py:31]: @jonavellecuerdo - I'm sorry I didn't propose this the first time around, but seeing it on a 2nd pass what do you think about putting `["gismit", "gisogm"]` into a config level constant?  Then we could refer to that constant (list) throughout the code; counting at least 3-4 instances where we check a variation of the source name being one of those two values.

File: lambdas/format_input.py [hunk 7]
```diff
@@ -29,6 +27,10 @@ def lambda_handler(event: dict, _context: dict) -> dict:
     }
 
     if next_step == "extract":
+        if source in ["gismit", "gisogm"]:
```
Comment (ghukill) [inline lambdas/format_input.py:30]: Example of where we could check against the constant list

File: lambdas/helpers.py [hunk 8]
```diff
@@ -54,7 +55,7 @@ def generate_step_output_filename(
     """
     sequence_suffix = f"_{sequence}" if sequence else ""
     if step == "extract":
-        file_type = "xml"
+        file_type = "jsonl" if source in ["gismit", "gisogm"] else "xml"
```
Comment (ghukill) [inline lambdas/helpers.py:58]: And again here, re: checking against a config level constant list.

**Review by ghukill — COMMENTED** [review]
Looking good.  Proposed one more change, define the source list `["gismit", "gisogm"]` as a config-level constasnt, and that can be checked against throughout the code, as opposed to checking against the hardcoded list.  And apologies again for not proposing that the first time around.

But, marking this PR as "Comment" because with or without that change, otherwise looks good and approved!  Please feel free to merge, from my POV, with or without that change.  

Logistically, once this is merged, I'd suggest we keep GDT-114 and GDT-115 open until we run the StepFunction through a few test runs now that most things are in place.  We can make sure these two stories are aligned with respect to the StepFunction flow.

**Comment by ghukill** [comment]
> Looking good. Proposed one more change, define the source list `["gismit", "gisogm"]` as a config-level constasnt, and that can be checked against throughout the code, as opposed to checking against the hardcoded list. And apologies again for not proposing that the first time around.
> 
> But, marking this PR as "Comment" because with or without that change, otherwise looks good and approved! Please feel free to merge, from my POV, with or without that change.
> 
> Logistically, once this is merged, I'd suggest we keep GDT-114 and GDT-115 open until we run the StepFunction through a few test runs now that most things are in place. We can make sure these two stories are aligned with respect to the StepFunction flow.

Looks as though the deployed version to Dev1 is operating as hoped!  

- [StepFunction execution for `gismit` / `daily`, uses the GeoHarvester with correct arguments](https://us-east-1.console.aws.amazon.com/states/home?region=us-east-1#/v2/executions/details/arn:aws:states:us-east-1:222053980223:execution:timdex-ingest-dev:fe3488b5-4e0c-47b2-b652-402702ef5021)
- [StepFunction execution for `libguides` / `daily` that correctly invokes the OAIHarvester](https://us-east-1.console.aws.amazon.com/states/home?region=us-east-1#/v2/executions/details/arn:aws:states:us-east-1:222053980223:execution:timdex-ingest-dev:fe3488b5-4e0c-47b2-b652-402702ef5021)

@jonavellecuerdo - if and when you merge this PR, from my POV, I think it's probably also safe to also mark [GDT-115](https://mitlibraries.atlassian.net/browse/GDT-115) as done.

File: lambdas/commands.py (hunk 6 above)
Comment (jonavellecuerdo) [reply lambdas/commands.py]: Definitely agree this is a good change to make!

**Comment by jonavellecuerdo** [comment]
@ghukill and @ehanson8 -- Please see the latest commit, addressing your comments. :)

File: lambdas/config.py [hunk 9]
```diff
@@ -1,18 +1,19 @@
 import logging
 import os
 
+GIS_SOURCES = ["gismit", "gisogm"]
 INDEX_ALIASES = {
     "rdi": ["jpal", "whoas", "zenodo"],
     "timdex": ["alma", "aspace", "dspace"],
-    "geo": ["gismit", "gisogm"],
+    "geo": GIS_SOURCES,
```
Comment (ghukill) [inline lambdas/config.py:8]: Looks great!  Thanks for making this change.

**Review by ehanson8 — APPROVED** [review]
Great final change!
